### PR TITLE
`Notifications`: Fix incorrect mail subjects

### PR DESCRIPTION
**Pull request by danielbode**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

### Motivation and Context

There was a bug where the email for some notification types did not have the correct subject but a placeholder.
### Description

For some notification types, the mail subject was not generated based on the user language and notification type, but was just taken from the notification title. This means previously to https://github.com/ls1intum/Artemis/pull/6106 the subject always was the English text and after https://github.com/ls1intum/Artemis/pull/6106 it was always set to a placeholder value. I added a German and English translation in case there was none. So now the mail subjects are displayed correctly and are translated.

### Steps for Testing

Prerequisites:
- Test on a test server that has set up a mail service (ts0 or ts1)
- Log in as a user whose emails you can access (your tum online account)
- 1 Instructor

1. Log in to Artemis
2. Click on your username -> settings -> notifications
Screenshot 2023-04-17 at 17-39-31 Notification Settings
3. Activate the email notifications you want to test
4. Do the action to trigger the notification (e.g. release an exercise)
5. Check your email and verify that the email subject is set correctly

#### Performance Review
- [ ] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
#### Manual Tests
- [x] Test 1

### Test Coverage

No changes.

### Screenshots

After:
Bildschirmfoto 2023-04-17 um 17 34 21

Before:
Bildschirmfoto 2023-04-17 um 17 34 45

### CODE REVIEW

File: src/main/resources/i18n/messages_de.properties [hunk 1]
```diff
@@ -94,7 +94,7 @@ email.notification.aux.difficulty.medium=Mittel
 email.notification.aux.difficulty.hard=Schwer
 
 # Plagiarism
-email.plagiarism.title=Neuer Plagiatsfall: �bung "{0}" im Kurs "{1}"
+email.plagiarism.title=Neuer Plagiatsfall: Übung "{0}" im Kurs "{1}"
```
Comment (pal03377) [inline src/main/resources/i18n/messages_de.properties:97]: Encoding broke

File: src/main/resources/i18n/messages_de.properties [hunk 2]
```diff
@@ -111,3 +111,26 @@ email.notification.title.tutorialGroup.updated=Die Übungsgruppe {0} im Kurs {1}
 email.notification.title.tutorialGroup.deleted=Die Übungsgruppe {0} im Kurs {1} wurde gelöscht.
 email.notification.title.tutorialGroup.assigned={2} wies dir die Leitung der Übungsgruppe {0} im Kurs {1} zu.
 email.notification.title.tutorialGroup.unassigned={2} entzog dir die Leitung der Übungsgruppe {0} im Kurs {1}.
+
+# Email Subjects
+# The reason for the format artemisApp.{notificationCategory}.title.{notificicationType} is that these placeholders are also used in the client and this is the format used there
+artemisApp.groupNotification.title.attachmentChange = Anhang aktualisiert
+artemisApp.groupNotification.title.exerciseReleased = Aufgabe veröffentlicht
```
Comment (pal03377) [inline src/main/resources/i18n/messages_de.properties:118]: Encoding broke

File: src/main/resources/i18n/messages_de.properties [hunk 3]
```diff
@@ -111,3 +111,26 @@ email.notification.title.tutorialGroup.updated=Die Übungsgruppe {0} im Kurs {1}
...
+artemisApp.groupNotification.title.exercisePractice = Aufgabe offen zum Üben
+artemisApp.groupNotification.title.duplicateTestCase = Doppelte Testfälle wurden gefunden.
```
Comment (pal03377) [inline src/main/resources/i18n/messages_de.properties:120]: Encoding broke

File: src/main/resources/i18n/messages_de.properties [hunk 4]
```diff
@@ -111,3 +111,26 @@ email.notification.title.tutorialGroup.updated=Die Übungsgruppe {0} im Kurs {1}
 email.notification.title.tutorialGroup.deleted=Die Übungsgruppe {0} im Kurs {1} wurde gelöscht.
 email.notification.title.tutorialGroup.assigned={2} wies dir die Leitung der Übungsgruppe {0} im Kurs {1} zu.
 email.notification.title.tutorialGroup.unassigned={2} entzog dir die Leitung der Übungsgruppe {0} im Kurs {1}.
+
+# Email Subjects
+# The reason for the format artemisApp.{notificationCategory}.title.{notificicationType} is that these placeholders are also used in the client and this is the format used there
+artemisApp.groupNotification.title.attachmentChange = Anhang aktualisiert
+artemisApp.groupNotification.title.exerciseReleased = Aufgabe veröffentlicht
+artemisApp.groupNotification.title.exercisePractice = Aufgabe offen zum Üben
```
Comment (pal03377) [inline src/main/resources/i18n/messages_de.properties:119]: Encoding broke

File: src/main/resources/i18n/messages_de.properties [hunk 5]
```diff
@@ -111,3 +111,26 @@ email.notification.title.tutorialGroup.updated=Die Übungsgruppe {0} im Kurs {1}
...
+artemisApp.groupNotification.title.duplicateTestCase = Doppelte Testfälle wurden gefunden.
+artemisApp.groupNotification.title.newAnnouncementPost = Neue Ankündigung
+
+artemisApp.singleUserNotification.title.exerciseSubmissionAssessed = Übungsabgabe bewertet
```
Comment (pal03377) [inline src/main/resources/i18n/messages_de.properties:123]: Encoding broke

File: src/main/resources/i18n/messages_de.properties [hunk 6]
```diff
@@ -111,3 +111,26 @@ email.notification.title.tutorialGroup.updated=Die Übungsgruppe {0} im Kurs {1}
...
+artemisApp.singleUserNotification.title.tutorialGroupRegistrationStudent = Du wurdest f�r eine Übungsgruppe registriert
+artemisApp.singleUserNotification.title.tutorialGroupDeregistrationStudent = Du wurdest von einer Übungsgruppe abgemeldet
+artemisApp.singleUserNotification.title.tutorialGroupRegistrationTutor = Ein Student wurde für deine Übungsgruppe registriert
+artemisApp.singleUserNotification.title.tutorialGroupDeregistrationTutor = Ein Student wurde von deiner Übungsgruppe abgemeldet
+artemisApp.singleUserNotification.title.tutorialGroupMultipleRegistrationTutor = Mehrere Studenten wurden zu deiner Übungsgruppe registriert
+artemisApp.singleUserNotification.title.tutorialGroupAssigned = Du wurdest zugeteilt eine Übungsgruppe zu leiten
+artemisApp.singleUserNotification.title.tutorialGroupUnassigned = Dir wurde die Leitung einer Übungsgruppe entzogen
+
+artemisApp.tutorialGroupNotification.title.tutorialGroupDeleted = Übungsgruppe gelöscht
+artemisApp.tutorialGroupNotification.title.tutorialGroupUpdated = Übungsgruppe aktualisiert
```
Comment (pal03377) [inline src/main/resources/i18n/messages_de.properties:136]: Encoding broke for all of these

File: src/main/webapp/i18n/de/notification.json [hunk 7]
```diff
@@ -77,8 +77,8 @@
...
                 "newManualFeedbackRequest": "Die Programmieraufgabe \"{{ placeholderValues.1 }}\" im Kurs \"{{ placeholderValues.0 }}\" hat eine neue manuelle Feedbackanfrage. Bitte bewerte das Feedback vor der Deadline.",
                 "illegalSubmission": "Die Aufgabe \"{{ placeholderValues.1 }}\" im Kurs \"{{ placeholderValues.0 }}\" hat illegale Abgaben von Studenten.",
-                "newExercisePost": "Die Aufgabe \"{{ placeholderValues.8 }}\" im Kurs \"{{ placeholderValues.0 }}\" hat einen neuen Beitrag: \"{{ placeholderValues.2 }}\"",
-                "newLecturePost": "Die Vorlesung \"{{ placeholderValues.8 }}\" im Kurs \"{{ placeholderValues.0 }}\" hat einen neuen Beitrag: \"{{ placeholderValues.2 }}\".",
+                "newExercisePost": "Die Aufgabe \"{{ placeholderValues.5 }}\" im Kurs \"{{ placeholderValues.0 }}\" hat einen neuen Beitrag: \"{{ placeholderValues.2 }}\"",
+                "newLecturePost": "Die Vorlesung \"{{ placeholderValues.5 }}\" im Kurs \"{{ placeholderValues.0 }}\" hat einen neuen Beitrag: \"{{ placeholderValues.2 }}\".",
```
Comment (pal03377) [inline src/main/webapp/i18n/de/notification.json:81]: Can you maybe explain that with a comment? Why do we access `placeholderValues.5` instead of some real key? 🤔

**Review by pal03377 — CHANGES_REQUESTED** [review]
The encoding seems to be broken in a lot of places :/

**Review by pal03377 — APPROVED** [review]
If the encoding for umlauts is fine like this (which it apparently is), the code is fine 👍

File: src/main/webapp/i18n/de/notification.json (hunk 7 above)
Comment (danielbode) [reply src/main/webapp/i18n/de/notification.json:81]: These are the texts for notifications. A notification does not contain the complete exercise or lecture object for example, but just an array of strings with placeholders. They are accessed based on an index.
There are lots of different types of notifications and they can have a lot of different placeholders. So if you would want an object with named attributes instead of a string array it would have to cover every possible placeholder value and most of them would be null. So we decided to use a string array. But this design decision was also not part of this PR.
I just noticed that we set the wrong index for these two notification texts so I fixed them in addition to the incorrect mail subject.

File: src/main/webapp/i18n/de/notification.json [hunk 8]
```diff
@@ -77,8 +77,8 @@
                 "programmingTestCasesChanged": "Die Testfälle der Programmieraufgabe \"{{ placeholderValues.1 }}\" im Kurs \"{{ placeholderValues.0 }}\" wurden aktualisiert. Die Abgaben der Studenten sollten erneut gebaut und getestet werden, um neue Ergebnisse zu erstellen.",
                 "newManualFeedbackRequest": "Die Programmieraufgabe \"{{ placeholderValues.1 }}\" im Kurs \"{{ placeholderValues.0 }}\" hat eine neue manuelle Feedbackanfrage. Bitte bewerte das Feedback vor der Deadline.",
                 "illegalSubmission": "Die Aufgabe \"{{ placeholderValues.1 }}\" im Kurs \"{{ placeholderValues.0 }}\" hat illegale Abgaben von Studenten.",
-                "newExercisePost": "Die Aufgabe \"{{ placeholderValues.8 }}\" im Kurs \"{{ placeholderValues.0 }}\" hat einen neuen Beitrag: \"{{ placeholderValues.2 }}\"",
```
Comment (TimOrtel) [inline src/main/webapp/i18n/de/notification.json:80]: What does this have to do with the PR? Also, why did you change this from value 8 to 5? Do we have to change this in the native application, too?

**Review by TimOrtel — COMMENTED** [review]
Code looks good. But I am not sure what effects this PR has.

File: src/main/webapp/i18n/de/notification.json (hunk 8 above)
Comment (danielbode) [reply src/main/webapp/i18n/de/notification.json:80]: It is not related to the incorrect email subject, but an additional bugfix. But as it is really small I also included it here. There is no value 8. 5 is the correct value. So with 8 the notification is not displayed correctly.
I don't think you have to adjust the mobile applications if you implemented it based on the list: https://confluence.ase.in.tum.de/display/ArTEMiS/Notifications. There it correctly says the exercise title is 5.

File: src/main/resources/i18n/messages_de.properties [hunk 9]
```diff
@@ -111,3 +111,26 @@ email.notification.title.tutorialGroup.updated=Die Übungsgruppe {0} im Kurs {1}
...
+artemisApp.groupNotification.title.exerciseReleased = Aufgabe veröffentlicht
+artemisApp.groupNotification.title.exercisePractice = Aufgabe offen zum Üben
+artemisApp.groupNotification.title.duplicateTestCase = Doppelte Testfälle wurden gefunden.
+artemisApp.groupNotification.title.newAnnouncementPost = Neue Ankündigung
+
+artemisApp.singleUserNotification.title.exerciseSubmissionAssessed = Übungsabgabe bewertet
+artemisApp.singleUserNotification.title.fileSubmissionSuccessful = Dateiabgabe erfolgreich
+artemisApp.singleUserNotification.title.newPlagiarismCaseStudent = Neuer Plagiatsfall
+artemisApp.singleUserNotification.title.plagiarismCaseVerdictStudent = Urteil zu deinem Plagiatsfall
+artemisApp.singleUserNotification.title.tutorialGroupRegistrationStudent = Du wurdest f�r eine Übungsgruppe registriert
```
Comment (DominikRemo) [inline src/main/resources/i18n/messages_de.properties:127]: "für" was not displayed correctly in the e-mail I got

**Review by DominikRemo — APPROVED** [review]
Tested on ts1. Got the e-mails without encoding problems 👍

**Review by pal03377 — APPROVED** [review]
re-approve code

**Review by TimOrtel — APPROVED** [review]
Looks good to me now.
